### DISCUSSION TITLE
How to decompress and write data from a .tgz file?

### DISCUSSION THREAD

**Discussion by owencraston** [tsolomko/SWCompression#41] [Q&A] [answered]
I am trying to decompress a `tgz` file and write it to a specified directory. Upon success we will return the location of the uncompressed data. The contents of the `tgz` file is a directory with different js files inside it so I am not sure if that complicates things. Here is what I have so far that is not working.

```swift
  func uncompressTgzFile(atPath path: String, toDirectory directory: String) throws -> String {
    print("uncompressTgzFile called with", path, directory)
    let tgzData = try Data(contentsOf: URL(fileURLWithPath: path))
    let decompressedData = try! SWCompression.GzipArchive.unarchive(archive: tgzData)
    print("uncompressTgzFile decompressedData", decompressedData)
    let toFilePath = URL(fileURLWithPath: directory)
    try! decompressedData.write(to: toFilePath)
    return toFilePath.path
  }
```
Here are the params...
`path "/Users/owencraston/Library/Developer/CoreSimulator/Devices/91A956BB-437B-42D2-AE67-DF25C6751A33/data/Containers/Data/Application/FC2E87B9-E2BA-4F60-BB4D-DF530A406B1E/Documents/archive.tgz"`
`directory "/Users/owencraston/Library/Developer/CoreSimulator/Devices/91A956BB-437B-42D2-AE67-DF25C6751A33/data/Containers/Data/Application/FC2E87B9-E2BA-4F60-BB4D-DF530A406B1E/Documents"`

I am getting the error 
``` Error Domain=NSCocoaErrorDomain Code=512 "The file “Documents” couldn’t be saved in the folder “5A7C2620-2D6F-4261-9A7E-77212FE6B2E9”." UserInfo={NSFilePath=/Users/owencraston/Library/Developer/CoreSimulator/Devices/91A956BB-437B-42D2-AE67-DF25C6751A33/data/Containers/Data/Application/5A7C2620-2D6F-4261-9A7E-77212FE6B2E9/Documents, NSUnderlyingError=0x6000011dc000 {Error Domain=NSPOSIXErrorDomain Code=21 "Is a directory"}}``` 

for the ` try! decompressedData.write(to: toFilePath)` call.

Here is a file you can test with...
[filsnap-2.3.12.tgz](https://github.com/tsolomko/SWCompression/files/10963424/filsnap-2.3.12.tgz)

**Comment by tsolomko** [comment, marked as answer]
The `tgz` file is actually `tar.gz` file, which means that this is a TAR archive that was afterwards compressed with GZip. So when you do this:
```swift
  let decompressedData = try! SWCompression.GzipArchive.unarchive(archive: tgzData)
```
you get a TAR archive that has to be further processed with either `TarContainer.open` or `TarReader`.

The error you're seeing is unrelated to the above: you are trying to write a file into an existing path, which is a directory. You have to specify a file name in the URL you are writing to. `Data.write(to:)` doesn't automatically append any file names to the path that was provided.
